### DISCUSSION TITLE
Folder Templates using placeholders

### DISCUSSION THREAD

**Discussion by tuxpeople** [SilentVoid13/Templater#741] [Help] [answered]
Hi

Is it possible to use `Folder Templates` with placeholder or wildcards?

Example folder structure:
```
customers/
    company-a/
           meetings/
    company-b/
           meetings/
    company-c/
           meetings/
```

Goal: If I create a new note in `customers/*/meetings` Templater should use my meeting notes template.

I already tested it using a `*` but it didn't work. Is that possible at all?

**Comment by welpdx** [comment, marked as answer]
### adding wildcard function to api
The wildcard is not a bad idea, but it would have to implemented to the api. 

### A Workaround
Here's a solution that you can do yourself to get what you want.

We can check the note's parent folder. If parent folder == "meetings", then apply template. This would mean that regardless of the root folder name, as long as a note is in a fold called "meetings" it would apply that template. This is a quicker solution though.

If you want a solution that really captures what you want, then we can check for the note's immediate parent folder's name as well as the note's parent's parent's parent's folder name.

**Reply by tuxpeople** [reply]
@welpdx Thanks a lot for your response! I had a similar idea, but with choosing manually, but your idea would be better :-) :

```javascript
<%*
let itsnow = tp.date.now("YYYY_MM_DD");

// What type do we want to create?
myTYPE = await tp.system.suggester(['Meeting Notes', 'Company Page', 'Project Page', 'People', 'Empty Note'], ['Meeting', 'Company', 'Project', 'People', 'Empty']);

// Maybe someone pressed cancel
if (myTYPE == null) { return; }

// select template
if (myTYPE == "Meeting") {
	myTEMPLATE = "customer-meeting-note";
} else if (myTYPE.startsWith("Company")) {
	myTEMPLATE = "customer-company";
} else if (myTYPE.startsWith("Project")) {
	myTEMPLATE = "customer-project";
} else if (myTYPE.startsWith("People")) {
	myTEMPLATE = "customer-people";
} else {
	myTEMPLATE = "customer-notes";
}

await tp.file.include("[[Templates/" + myTEMPLATE + "]]");
%>
```


I struggled with the variables of the included template, and also with moving those files around.

Not being very good at Javascript, I assume something like this would implement your suggestion:
```javascript
if (tp.file.path(true).endsWith("Meetings/")) { 
  // Meetings code
}
```

I'll try that. But first I need to understand how those variables works in included templates. I might end up doing all the logic in that meta-template and just include a template with output of a few variables here and there.

**Reply by tuxpeople** [reply]
@welpdx The following is working almost as expected. 

I know, my extensive usage of `startsWith` is not necessary, but that's the only kind of comparison I got working with my very limited JavaScript knowhow.

The only thing not working now is the `tp.file.include` part. Which is a big issue, as it makes the whole thing useless ;-) it just silently includes nothing. I'll search discussion for help on this and will otherwise open a new discussion, as it's a different problem.

```javascript
<%*
itsnow = tp.date.now("YYYY_MM_DD");
language = `de`;
title = tp.file.title  
rootfolder = "Work"

// Get folders relative path
filepath = tp.file.path(true) || 'abc';
filepath_root = filepath.split('/')[1] || 'abc';
filepath_scope = filepath.split('/')[2] || 'abc';
filepath_type = filepath.split('/')[3] || 'abc';

// Is it for a customer or us?
scope = filepath_root;
if (!(scope == 'Company' || scope == 'Customers')) {
  scope = await tp.system.suggester(["Company", "Customer"], ["Company", "Customers"]);
}
if (scope == "Company") {
  basepath = `${rootfolder}/${scope}/`
} else {
  if (!(await this.app.vault.adapter.exists(`${rootfolder}/${scope}/`))) {
      await this.app.vault.createFolder(`${rootfolder}/${scope}/`);
  }
  const subPathList = ['-New-'];
  // Read folders from customers path
  let pathList = await this.app.vault.adapter.list(`${rootfolder}/${scope}/`);
  for( d of pathList.folders ) {
    d1 = d.substring(d.lastIndexOf('/') + 1);
    d2 = d1.split('.')[0];
    subPathList.push(`${d2}`);
  }
  if (!(subPathList.includes(filepath_scope))) {
    // Select a customer
    customer = await tp.system.suggester(subPathList, subPathList);
    // Is it a new customer?
    if (customer.startsWith('-New-')) {
      customer = await tp.system.prompt("Customer");
    }
  } else {
    customer = filepath_scope;
  }
  basepath = `${rootfolder}/${scope}/${customer}/`
}

// What are we aiming for?
if (filepath_type.startsWith("Meetings")) { 
  type = "Meetings";
} else if (filepath_type.startsWith("Company")) { 
  type = "Companies";
} else if (filepath_type.startsWith("Project")) { 
  type = "Projects";
} else if (filepath_type.startsWith("People")) { 
  type = "People";
} else if (filepath_type.startsWith("Notes")) { 
  type = "Notes";
} else {
  type = await tp.system.suggester(['Meeting Notes', 'Company Page', 'Project Page', 'People', 'Empty Note'], ['Meetings', 'Companies', 'Projects', 'People', 'Notes']);
}

// What do we need to know?
if ( type.startsWith("Meetings")) { 
  topic = await tp.system.prompt("Topic");
  title = `Meeting Notes - ${itsnow} - ${topic}`;
  workdir = `${basepath}/${type}/`;
} else if ( type.startsWith("Companies")) { 
  title = `${customer}`;
  workdir = `${basepath}/`;
} else if ( type.startsWith("Projects")) { 
  project = await tp.system.prompt("Project");
  title = `${project}`;
  workdir = `${basepath}/${type}/`;
} else if ( type.startsWith("People")) { 
  name = await tp.system.prompt("Name of the person");
  title = `${name}`;
  workdir = `${basepath}/${type}/`;
} else if ( type.startsWith("Notes")) { 
  topic = await tp.system.prompt("Topic");
  title = `Note - ${topic} - ` + tp.file.creation_date("YYYY-MM-DD-HH-mm");
  workdir = `${basepath}/${type}/`;
}

if (!(await this.app.vault.adapter.exists(workdir))) {
    await this.app.vault.createFolder(workdir);
}

// if (title.startsWith("Untitled")) {
//     title = await tp.system.prompt("Title");
// }

await tp.file.move(workdir + title);  

await tp.file.include(`[[fragment-${type}]]`);

%>
```

**Reply by welpdx** [reply]
Wow! Great work! Sorry I am so late on my reply. I must have missed this.

Anyways, in regards to the include, you got the code right. You are now just missing the part where you get the content and output it to the note at the very end. 

To do that, just add `tR+=` before the last line. So that it looks like this:
```js
tR+= await tp.file.include(`[[fragment-${type}]]`);
```

You can read more about tR and what it does [here](https://silentvoid13.github.io/Templater/commands/execution-command.html?highlight=tR#how-to-output-a-value-from-a-javascript-execution-command-).

**Reply by welpdx** [reply]
ok nvm, ignore above. I just realized that we got this over at another discussion (#805) lmao. mb
